http_request.go

```diff
@@ -15,10 +17,35 @@ func Do(req *http.Request, env Environment) (string, error) {
 	if env.HTTPClient != nil {
 		client = env.HTTPClient
 	}
-	response, err := client.Do(req)
+
+	backoffConfig := env.BackoffConfig
+	if backoffConfig == nil { // if no backoff configured, use default one
+		backoffConfig = backoff.NewExponentialBackOff()
+	}
+
+	if req.Method != http.MethodGet { // if the request is anything other than GET, don't use backoff
+		backoffConfig = &backoff.StopBackOff{}
+	}
+
+	var response *http.Response
+	var err error
+	// Retry on HTTP 429 rate limit, or network error, see:
+	// https://godoc.org/github.com/cenkalti/backoff#pkg-constants
+	backoff.Retry(func() error {
+		response, err = client.Do(req)
+		if err != nil {
+			return nil
+		}
+		if response == nil || response.StatusCode == http.StatusTooManyRequests {
```

Comment:
Are you implying network error from `response== nil`?

Reply:
You are right, we should instead retry if there _is_ `err`.
> A non-2xx response doesn’t cause an error.
But, keep it saved to return after last retry.

We took it from here, but not perfectly, https://github.com/chartmogul/chartmogul-go/blob/be00f9b814ec5d1ab64fe2e4ff88e3e8df850718/generic.go#L29

Reply:
You can borrow from here: https://github.com/chartmogul/go-utilities/blob/a9ddca6f22358acdc1fbeeb38837ac578de81559/utils/errors.go#L112